packages/connect-react/src/hooks/form-context.tsx

```diff
@@ -395,6 +436,98 @@ export const FormContextProvider = <T extends ConfigurableProps>({
+    let newErrors = [
+      ...sdkErrors,
+    ]
+    if (!o) return
+    const handleObservationErrors = (observation: never) => {
+      const name: string = observation.err?.name
+      const message: string = observation.err?.message
+      if (name && message) return {
+        name,
+        message,
+      } as Record<string, string>
+    }
+
+    if (Array.isArray(o) && o.length > 0) {
+      for (let i = 0; i < o.length; i++) {
+        const item = o[i]
+        if (typeof item === "string") {
+          try {
+            const json = JSON.parse(item)
+            const name = json.name
+            const message = json.message
+            if (name && message) {
+              newErrors.push({
+                name,
+                message,
+              } as Record<string, string>)
+            }
+            // eslint-disable-next-line @typescript-eslint/no-unused-vars
+          } catch (e) {
+            // pass
+          }
+        } else if (typeof item === "object" && "name" in item && "message" in item) {
+          const name = item.name
+          const message = item.message
+          if (name && message) {
+            newErrors.push({
+              name,
+              message,
+            } as Record<string, string>)
+          }
+        } else if (typeof item === "object" && item.k === "error") {
+          const res = handleObservationErrors(item)
+          if (res) newErrors.push(res)
+        }
+      }
+    } else if (typeof o === "object" && "os" in o || "observations" in o) {
+      const os = o.os || o.observations
+      if (Array.isArray(os) && os.length > 0) {
+        newErrors.push(
+          ...os.filter((it) => it.k === "error")
+            .map(handleObservationErrors)
+            .filter((e) => e !== undefined),
+        )
+      }
+    } else if (typeof o === "object" && "message" in o) {
+      // Handle HTTP errors thrown by the SDK
+      try {
+        const json = JSON.parse(o.message)
+        const data = json.data
+        if (data && "observations" in data) {
+          const obs = data.observations || []
+          if (obs && obs.length > 0) {
+            const res = obs.filter((it) => it.k === "error")
+              ?.map(handleObservationErrors)
+              ?.filter((e) => e !== undefined) || []
+            newErrors.push(
+              ...res,
+            )
+          }
+        } else if (data && "error" in data && "details" in data) {
+          newErrors.push({
+            name: data.error,
+            message: JSON.stringify(data.details),
+            //     message: ` // TODO: It would be nice to render the JSON in markdown
+            // \`\`\`json
+            // ${JSON.stringify(data.details)}
+            // \`\`\`
+            // `,
+            //   })
+          })
+        }
+        // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      } catch (e) {
+        // pass
+      }
+    } else {
```

Comment:
I'd put each one of these cases in their own functions, it's kind of hard to follow since these condition checks are also cryptic.